apps/req.c

```diff
@@ -1744,18 +1744,23 @@ static int do_sign_init(EVP_MD_CTX *ctx, EVP_PKEY *pkey,
 int do_X509_sign(X509 *x, EVP_PKEY *pkey, const EVP_MD *md,
                  STACK_OF(OPENSSL_STRING) *sigopts)
 {
-    int rv;
+    int rv, try_free = 0;
     EVP_MD_CTX *mctx = EVP_MD_CTX_new();
 #ifndef OPENSSL_NO_SM2
     EVP_PKEY_CTX *pctx = NULL;
 #endif
 
     rv = do_sign_init(mctx, pkey, md, sigopts);
-    if (rv > 0)
+    if (rv > 0) {
         rv = X509_sign_ctx(x, mctx);
+        try_free = 1;
+    }
 #ifndef OPENSSL_NO_SM2
-    /* only in SM2 case we need to free the pctx explicitly */
-    if (ec_pkey_is_sm2(pkey)) {
+    /*
```

Comment:
Probably better not to use try_free variable as it may complain if OPENSSL_NO_SM2 is defined.
Move the code up instead ie. 
if (rv > 0) {
   rv = X509_sign_ctx(x, mctx);
#ifndef OPENSSL_NO_SM2
#endif
}

Reply:
Sure, no problem